internal/metrics/metrics.go

```diff
@@ -1,32 +1,58 @@
 package metrics
 
 import (
+	"time"
+
 	"github.com/RHEnVision/provisioning-backend/internal/kafka"
 	"github.com/RHEnVision/provisioning-backend/internal/models"
+	"github.com/RHEnVision/provisioning-backend/internal/version"
 	"github.com/prometheus/client_golang/prometheus"
 )
 
 const (
-	metricNameSourceAvailabilityCheckRequestTotal = "source_availability_check_request_total"
+	metricNameSourceAvailabilityCheckRequestTotal    = "source_availability_check_request_total"
+	metricNameSourceAvailabilityCheckRequestDuration = "source_availability_check_request_duration_ms"
 )
 
 var TotalAvailabilityCheckReqs = prometheus.NewCounterVec(
 	prometheus.CounterOpts{
 		Name:        metricNameSourceAvailabilityCheckRequestTotal,
 		Help:        "availability Check requests count partitioned by type (aws/gcp/azure), source status, component and error",
-		ConstLabels: prometheus.Labels{"service": "provisioning"},
+		ConstLabels: prometheus.Labels{"service": version.PrometheusLabelName},
 	},
 	[]string{"type", "status", "component", "error"},
 )
 
-func IncTotalAvailablilityCheckReqs(provider models.ProviderType, component string, StatusType kafka.StatusType, err error) {
+var AvailabilityCheckReqsDuration = prometheus.NewHistogramVec(
+	prometheus.HistogramOpts{
+		Name:        metricNameSourceAvailabilityCheckRequestDuration,
+		Help:        "Availability check Request duration partitioned by type and error",
+		ConstLabels: prometheus.Labels{"service": version.PrometheusLabelName},
+	},
+	[]string{"type", "error"},
+)
+
+func ObserveAvailablilityCheckReqsDuration(provider models.ProviderType, ObservedFunc func() error) {
+	start := time.Now()
+	err := ObservedFunc()
+	errString := "false"
+	if err != nil {
+		errString = "true"
+	}
+
+	defer func() {
+		AvailabilityCheckReqsDuration.WithLabelValues(provider.String(), errString).Observe(float64(time.Since(start).Nanoseconds()) / 1000000)
+	}()
```

Comment:
You need to move the defer call right behind `time.Now()` call otherwise it will not be used when error occurs. The correct ordering is:

```go
errString := "false"
start := time.Now()
defer func() { ... }

err := ObservedFunc()
if err != nil {
	errString = "true"
}
```

Reply:
No it is deferred execution, it will be _registered_ for execution before the func call, but _executed_ only and only the function (where `defer` is present) exists (e.g. via regular code path or panic).

Reply:
Thank you lzap. 
I think that these examples you shared with me emphasize the need here of an anonymous function: 
https://go.dev/play/p/GFdfLjoYK1X
https://go.dev/play/p/UWBX8324Z9U